### ISSUE TITLE
Improvement of Physics

### ISSUE THREAD

**Issue by KarlZeilhofer** [juzzlin/DustRacing2D#37]
Hi,
great Game, just played it half an hour! :+1: 
Thank you very much for providing such a great content. 

Is there anything we can do about the physics of the car? It feels like driving a front motor car with rear wheels drive and 70/30 percent weight distribution between front and rear. 

Would love to have a 4 wheel drive with balanced weight distribution :)

Greets, Karl

(v2.0.1 from the ubuntu repo on linux mint)

**Comment by juzzlin** [comment]
I'm glad to hear you liked the game :)

The behavior of the car is mostly defined by the tire models (tire.hpp/tire.cpp) initialized in the constructor of the car class (car.cpp). I haven't been working on this project for a while, but I think it could be nice to load these car parameters from a JSON or something. I might actually implement that. At some point I didn't know anymore what is good and what is bad (you become blind to your own game) so it would be nice to have someone to give his opinion of the balance of the car etc.

**Comment by juzzlin** [comment]
I'll accept if you use XML or JSON :) QSettings is kinda deprecated and not the best option for this kind of stuff in my opinion. The game already uses QtXml for tracks. For JSON this is like magic: https://github.com/nlohmann/json and I prefer it Qt's JSON implementation. That could be just added to src/game/contrib/nlohmann/ or something like that. IMO the most easy-to-use solution as the API is just like std::map.

**Comment by juzzlin** [comment]
For the default parameters we could have something like:

```
{
  "power": 5000,
  "mass": 1500,
  "frontFriction": 0.85,
  "rearFriction": 0.95,
  ...
}

```

Maybe even tire locations. 

I also noticed that some basic parameters are in car.hpp in Car::Description. This would indeed need some refactoring :)

**Comment by KarlZeilhofer** [comment]
Indeed, the json lib from nlohmann looks very impressive. But currently I'm stuck on compiling the game at all, see #39

**Comment by KarlZeilhofer** [comment]
I'm just collecting my suggestions and thoughts here: 

## Inertia:
`physicsComponent().setMomentOfInertia(desc.mass * 3.0);`  
for a rectangle, the inertia is 1/12*(a^2+b^2)*m. For a car I suggest 1.4x4m, which gives about 1.5*m instead of programmed 3*m. This is the inertia for the rotation about the center of gravity. Translated to the rear axle, this gives about 3.75*m, if the axle is 1.5m behind the center of gravity. 

Suggested value is 1.5, as it feels a lot better. 

@juzzlin Do you know, what center of rotation is assumed for that inertia?

**Comment by juzzlin** [comment]
The center of rotation is assumed to be the center of the object. Of course that's not entirely correct as many other things. 

According to the documentation in PhysicsComponent header the angular "impulse" is supposed to be rad/s. I did this many years ago so I don't remember anymore why is that. Seems to be incorrect.

The tire locations are partly like that because they should also match the graphics. The units are "scene units" and I remember there's a conversion factor somewhere. Yeah...seems to be in MCWorld. There's metersPerUnit() which is then used in some calculations. I had plans to get rid of this but didn't have enough time. 

The speed calculation has a magic number to make the value look "better" in the game. The raw speed _should_ be in m/s. I hope it's not some messed up "scene units" / s :)

I like your observations and appreciate all improvements and corrections to the physics calculations.

**Comment by juzzlin** [comment]
One more thing in the case you haven't noticed. You can run some physics framework tests in the build dir by:
`$ ctest`

It starts the CTest test runner and runs some unit tests that are done with Qt's test framework.

**Comment by juzzlin** [comment]
I used `git grep metersPerUnit` to find out the value is being used in MCImpulseGenerator. It would be cool if everything was in meters.

You are right about the braking distance. It's way too powerful and also goes to reverse too easily. I think that it was needed so that the computer players could brake hard enough. Of course that could be handled also in some other way like making it realistic for the human players.

**Comment by juzzlin** [comment]
The linear velocity calculations are in m/s, because it depends only on mass, acceleration etc. But the projection to the game scene and its dimensions is another thing :)

**Comment by KarlZeilhofer** [comment]
@juzzlin 
one question: Any suggestions, where I should do the calculations between scene coordinates and the physical coordinates? 

I have pushed now all my in-code comments, if you would like to have a look: https://github.com/KarlZeilhofer/DustRacing2D/tree/devs/physics  
I'm using the Qt Creators TODO plugin, to locate that comments easily. But you could also grep for `TODO` and `FIXME`. 

As a first step, I would like to do following things:
* define a proper scene scale, so the car is about 4.2m long (similar to BMW 1er or WV Golf). 
* keep the physic units in SI, it's mostly the length dimension and it's derivatives. 
* calculate correct force with physical speed
* correct tire friction for braking with 25% of car's mass (instead of 50%), since we have 4 tires. 
* calculate inertia on the fly by the given dimensions of the car with the formular `1/12 * (a² + b²)`
* perhaps add a enum for rear drive, front drive and all wheel drive. 
* try to play with the tires locations `MCVector3dF Car::leftFrontTireLocation()` and corresponding functions. 
* rename various `impulses` to force or momentum.

**Comment by juzzlin** [comment]
Looks like a good list. 

Maybe it'd be the "best" to call the conversion functions in MCWorld on-the-fly wherever needed..? Of course the perfect solution would be to have everything in SI (except rendering stuff in renderers, view objects and scene) but I guess that would be too big of a change for now :)

**Comment by KarlZeilhofer** [comment]
You are right with the velocity. But the transformation is needed somewhere. 

In first place, I'll try to improve the physics as they are. One big thing I understood now, is that the tires only generate forces, when they already slide. That algorithm seems to be too simple/unstable - somehow. But describing the system isn't an easy task, I think - though I once learned that thing at the university. If you have Qt installed, you may try to run https://github.com/KarlZeilhofer/inverse-tripple-pendulum  
It's also a sort of a game :smile:

**Comment by juzzlin** [comment]
I tried your app and it worked very well. I just wish I understood the math :)

The tire model uses a "trick" I learned somewhere to limit and lose grip and make it slide:

`impulse.clampFast(parent().physicsComponent().mass() * 7.0f * m_car.tireWearFactor());`

**Comment by juzzlin** [comment]
@KarlZeilhofer Are you still working on this?

**Comment by KarlZeilhofer** [comment]
Not really. I couldn't manage to convert everything to metric units, and a realistic physics engine is also not easy to implement. Though I've tweaked some values, and in my opinion it is now more fun to play. 

I've published my current WIP in commit adfd59a3c093a4e7e7e305221a0c8d3af3a18d6b

**Comment by juzzlin** [comment]
Would it be possible to make a cleaned-up pull request so it would be easier to see the changes?
